sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/action/ReflectiveActionRouter.scala

```diff
@@ -41,9 +42,25 @@ class ReflectiveActionRouter[A <: Action](action: A, componentMethods: Map[Strin
         componentMethod.requestMessageDescriptor,
         message.metadata())
 
-    componentMethod.method.get // safe call: if component method is None, proxy won't forward calls to it
-      .invoke(action, componentMethod.parameterExtractors.map(e => e.extract(context)): _*)
-      .asInstanceOf[Action.Effect[_]]
+    val inputTypeUrl =
+      message.payload().asInstanceOf[ScalaPbAny].typeUrl.replace(kalix.javasdk.JsonSupport.KALIX_JSON, "")
+
+    if (componentMethod.methodsMap.isEmpty)
+      componentMethod.method.get // safe call: if component is None the proxy won't forward calls to it
+        .invoke(action, componentMethod.parameterExtractors.map(e => e.extract(context)): _*)
+        .asInstanceOf[Action.Effect[_]]
+    else {
+      val method = Option(componentMethod.methodsMap.getOrElse(inputTypeUrl, null))
+      method.get // safe call: if component is None the proxy won't forward calls to it
+        .invoke(
+          action,
+          componentMethod.parameterExtractors.collect {
+            case e @ AnyBodyExtractor(cls)
+                if cls.getName.equals(inputTypeUrl) => //TODO This need to be tested thoroughly
+              e.extract(context)
```

Comment:
This is needed because it enters this line for each one of the extractors. We have two extractors here, for example in the test we would have ValueIncreased and ValueMultiplied extractors. Here is where we are saying only pick the relevant extractor. Makes sense?

Reply:
Ok, I see. 

But in that case, we should not use `cls.getName` to compare with the `typeUrl`. The `typeUrl` will hopefully not be the class name, but whatever the user defined in the `@JsonTypeInfo`.

One approach we could try here is to not extends the ComponentMethod with a Map, but use a ComponentMethod composite. If we move the logic to lookup the method to the ComponentMethod, then the composite will always return the right one and we will automatically have the right extractor.

Reply:
fixed now and moved the logic to component method